runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -5118,172 +5118,260 @@ TR_PrexArgInfo* TR_PrexArgInfo::argInfoFromCaller(TR::Node* callNode, TR_PrexArg
+   auto prexArgInfo = TR_PrexArgInfo::enhance(prexArgInfoFromTarget, prexArgInfoFromCallSite, comp());
+
+   if (tracePrex && prexArgInfo)
       {
-      if (tracePrex)
-         traceMsg(comp(), "PREX.inl: Propagating prex argInfo from caller for [%p] %s %s\n",
-                  site->_callNode,
-                  site->_callNode->getOpCode().getName(),
-                  site->_callNode->getSymbol()->castToMethodSymbol()->getMethod()->signature(trMemory(), stackAlloc));
+      traceMsg(comp(), "PREX.inl:    argInfo for target %p\n", target);
+      prexArgInfo->dumpTrace();
+      }
 
-      TR_PrexArgInfo* argsFromCaller = TR_PrexArgInfo::argInfoFromCaller(site->_callNode, callerArgInfo);
-      target->_prexArgInfo = TR_PrexArgInfo::enhance(target->_prexArgInfo, argsFromCaller, comp());
+   // At this stage, we can improve the type of the virtual guard we are going to use
+   // For a non-overridden guard or for an interface guard, if it makes sense, try to use a vft-test
+   // based virtual guard
+   //
+   bool disablePreexForChangedGuard = false;
+   TR_PersistentCHTable * chTable = comp()->getPersistentInfo()->getPersistentCHTable();
+   TR_PersistentClassInfo *thisClassInfo = chTable->findClassInfoAfterLocking(target->_receiverClass, comp());
+   if (target->_calleeSymbol->hasThisCalls() &&
+       target->_receiverClass &&
+       !TR::Compiler->cls.isAbstractClass(comp(), target->_receiverClass) &&
+       !fe()->classHasBeenExtended(target->_receiverClass) &&
+       thisClassInfo &&
+       thisClassInfo->isInitialized() &&
+       ((target->_guard->_kind == TR_NonoverriddenGuard && target->_guard->_type == TR_NonoverriddenTest) ||
+        (target->_guard->_kind == TR_InterfaceGuard)) &&
+       performTransformation(comp(), "O^O VIRTUAL GUARD IMPROVE: Changed guard kind %s type %s to use VFT test\n", tracer()->getGuardKindString(target->_guard), tracer()->getGuardTypeString(target->_guard)))
+      {
+      target->_guard->_type = TR_VftTest;
+      target->_guard->_thisClass = target->_receiverClass;
       }
 
-   return computePrexInfo(target);
+   return prexArgInfo;
    }
 
 TR_PrexArgInfo *
 TR_J9InlinerUtil::computePrexInfo(TR_CallTarget *target)
    {
-   TR_CallSite *site = target->_myCallSite;
-   if (!site)
+   return computePrexInfo(target, NULL);
+   }
+
+/** \brief
+ *     Find the def to an auto or parm before treetop in a extended basic block
+ *
+ *  \return
+ *     The treetop containing the def (the store)
+ */
+static TR::TreeTop*
+defToAutoOrParmInEBB(TR::Compilation* comp, TR::TreeTop* treetop, TR::SymbolReference* symRef, TR::Node** valueNode)
+   {
+   while (treetop)
+      {
+      auto ttNode = treetop->getNode();
+      if (ttNode->getOpCodeValue() == TR::BBStart)
+         {
+         auto block = ttNode->getBlock();
+         if (!block->isExtensionOfPreviousBlock())
+            return NULL;
+         }
+
+      if (ttNode->getOpCode().isStoreDirect() &&
```

Comment:
You may want to consider using the getStoreNode api to get a store node even if it is under the treetop. That api gets you the store node if its the top tree or under a treetop and that in turn means you don't have to write code to check for those possibilities, while at the same time avoiding a functional bug where you may miss certain stores if they happen to be under a treetop for some reason.